### ISSUE TITLE
cal_AUC_ml_res

### ISSUE THREAD

**Issue by weifuq** [l-magnificence/Mime#22]
> all.auc.3y <- cal_AUC_ml_res(res.by.ML.Dev.Prog.Sig = res,train_data = mm[["Training_Dataset"]],
+                              inputmatrix.list = mm,mode = 'all',AUC_time = 3,
+                              auc_cal_method="KM")
--- Data preprocessing ---
[1] "Training_Dataset"
[1] "Testing_Dataset1"
[1] 2832.4
[1] "Testing_Dataset2"
[1] 2022
[1] "Testing_Dataset3"
[1] "RSF"
Error in survdiff.fit(y, groups, strata.keep, rho) : 
  There is only 1 group
  是因为分不出来组吗？
  生存分析的时候可以出结果
  尝试了mode = 'double'也是一样的错误
  all.auc.1y <- cal_AUC_ml_res(res.by.ML.Dev.Prog.Sig = res,train_data = mm[["Training_Dataset"]],
                             inputmatrix.list = mm,mode = 'double',AUC_time = 1,double_ml1 = "StepCox[both]",double_ml2 ="SuperPC",
                             auc_cal_method="KM")

**Comment by l-magnificence** [comment]
Can you show your riskscore based on RSF model?

**Comment by weifuq** [comment]
![image](https://github.com/l-magnificence/Mime/assets/154768214/f226b16b-64e0-4558-85ef-2460509d9344)

**Comment by l-magnificence** [comment]
Your OS.time is day unit?

**Comment by weifuq** [comment]
yes

**Comment by weifuq** [comment]
执行cal_AUC_ml_res函数的时候是不是要在list_train_vali_Data中加上分组信息？

**Comment by l-magnificence** [comment]
No need.
No such problems have been encountered.

**Comment by weifuq** [comment]
> all.auc.3y <- cal_AUC_ml_res(res.by.ML.Dev.Prog.Sig = res,train_data = mm[["Training_Dataset"]],
+                              inputmatrix.list = mm,mode = 'all',AUC_time = 3,
+                              auc_cal_method="KM")
--- Data preprocessing ---
[1] "Training_Dataset"
[1] "Testing_Dataset1"
[1] 2832.4
[1] "Testing_Dataset2"
[1] 2022
[1] "Testing_Dataset3"
[1] "RSF"
Error in survdiff.fit(y, groups, strata.keep, rho) : 
  There is only 1 group
  Error is why？

**Comment by kongwh99** [comment]
![image](https://github.com/user-attachments/assets/d8e7b5f4-c16b-4498-b2db-cce22c552a81)
I have encountered the same problem, can you solve this problem?

**Comment by weifuq** [comment]
No发自我的手机-------- 原始邮件 --------发件人： kongwh99 ***@***.***>日期： 2024年7月16日周二 下午1:41收件人： l-magnificence/Mime ***@***.***>抄送： weifuq ***@***.***>, Author ***@***.***>主    题： Re: [l-magnificence/Mime] cal_AUC_ml_res (Issue #22)
image.png (view on web)
I have encountered the same problem, can you solve this problem?

**Comment by l-magnificence** [comment]
Sorry, currently I don't know how to solve it due to no happening in my system. Anyone knows?

**Comment by csu-lzj** [comment]
I had the same problem,Can i calculate it like this?
time_roc_train <- timeROC(
  T = train_risk$OS.time,  
  delta = train_risk$OS,  
  marker = train_risk$RS,  
  cause = 1,  
  weighting="marginal", 
  times = c(365, 1095, 1825),
  ROC = TRUE,
  iid = TRUE  
)

**Comment by l-magnificence** [comment]
Yes, you can calculate AUC yourself using the results output by ML.Dev.Prog.Sig. Like this:

```
 risk.survivalROC <- survivalROC(
                Stime = x$OS.time,
                status = x$OS,
                marker = x$RS,
                predict.time = 365 * 1, 
                method = "KM"
              )
```

**Comment by kimyoki123** [comment]
我遇到了同样的问题，请问有解决办法了吗。 Error in survdiff.fit(y, groups, strata.keep, rho) :
There is only 1 group
Error is why？

**Comment by 123wanguang123456** [comment]
我遇到了同样的问题，请问有解决办法了吗。设置os.time 大于10,仍然解决不了
 Error in survdiff.fit(y, groups, strata.keep, rho) :
There is only 1 group
Error is why？

**Comment by JiahaoWongg** [comment]
@l-magnificence 有些算法得到的风险评分在所有样本中完全一样，导致无法分组，计算风险评分的过程是否有bug？

**Comment by JiahaoWongg** [comment]
@JiahaoWongg 但是这种情况是少数的模型和数据集，勉强的解决办法是给风险评分只有一个值的数据集的风险评分添加0.001~0.01倍的随机数值，这样就能正常分组了，且不会太影响检验结果，大概是下面这样实现的，在cal_AUC_ml_res.R的几个returnRStoROC字段前面加入下面的代码：

```R
datasets = names(rs)
for(dataset in datasets){
  risk_score_tb = rs[[dataset]]
  if(length(unique(risk_score_tb$RS)) == 1){
    rs_real = risk_score_tb$RS
    rs_fix  = rs_real + runif(length(rs_real), min = 0.001, max = 0.01) * rs_real
    rs[[dataset]]$RS = rs_fix
  }
}

```

可以`source`一下我改的函数看看：https://github.com/JiahaoWongg/Files/blob/main/cal_AUC_ml_res2.R

函数名为：`cal_AUC_ml_res2`

目前是能正常出图就行😂

<img width="911" height="1045" alt="Image" src="https://github.com/user-attachments/assets/42c26fcd-a465-4fb4-abbd-61e9b5e391bd" />
